### PR TITLE
Add task manager and collector

### PR DESCRIPTION
**Pull request by embano1**
## Description

This change introduces a task manager (https://vdc-download.vmware.com/vmwb-repository/dcr-public/8946c1b6-2861-4c12-a45f-f14ae0d3b1b9/a5b8094c-c222-4307-9399-3b606a04af55/vim.TaskManager.html) with support for creating a task history collector.

History collector is moved from `object` into its own package and changes were made to the event package accordingly (collector logic is shared between task and event manager).

Notes:

- followed event manager/collector approach
- won't work against vcsim as the task manager is not implemented there yet (#2567)
- moved `HistoryCollector` from `object` to its own package to avoid cyclic imports
- `CreateTask` method not implemented (not sure if this is useful), can be added later if need to
- `newHistoryCollector` not exported since it's merely a helper and does not add value to external API IMHO
  - did same for creating the history collector in the events package 

⚠️ **BREAKING** changes:

`event.Manager` does not embed `object.Common` anymore. Only the methods `Client()` and `Reference()` were implemented. `event.NewHistoryCollector()` is now unexported (`newHistoryCollector()`) as it was merely a helper and to comply with the task manager implementation.

Closes: #2497
Signed-off-by: Michael Gasch <mgasch@vmware.com>

## Type of change

Please mark options that are relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Tested example against vCenter 6.7.

**Test Configuration**:
* Toolchain:
* SDK:
* (add more if needed)

## Checklist:

- [x] My code follows the CONTRIBUTION [guidelines](./../CONTRIBUTING.md) of
  this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged

### CODE REVIEW

**Review by dougm — COMMENTED** [review]
Thanks @embano1 , +1 for adding this. We can probably find a way around the import issue, I'll try to take a look at that soon.  I can also help with the vcsim support

**Comment by embano1** [comment]
That'd be great Doug. No rush, I'm on PTO and will get back to this in a couple of weeks. Pretty sure you have found a workaround by then :)

**Comment by embano1** [comment]
TODOs:

- [ ] Create issue to add TaskManager in vcsim
- [ ] Add example for TaskManager

File: history/collector.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		This: c.Reference(),
+	}
+
+	_, err := methods.DestroyCollector(ctx, c.c, &req)
+	return err
+}
+
+func (c Collector) Reset(ctx context.Context) error {
+	req := types.ResetCollector{
+		This: c.Reference(),
```
Comment (embano1) [inline history/collector.go:67]: c.ref

**Comment by embano1** [comment]
@dougm ready for review. Updated PR notes incl. breaking changes related to the events package.

Something odd I found is that while the EventHistoryCollector returns items sorted in ascending order (by eventID), TaskHistoryCollector returns tasks sorted in descending order (by taskID). I don't see any code differences between the events and task implementation or how to influence this. So pretty sure this is server behavior then?

**Comment by dougm** [comment]
> Something odd I found is that while the EventHistoryCollector returns items sorted in ascending order (by eventID), TaskHistoryCollector returns tasks sorted in descending order (by taskID). I don't see any code differences between the events and task implementation or how to influence this. So pretty sure this is server behavior then?

Not sure, I've yet to try TaskHistoryCollector, but yes sounds like it's server behavior.

**Review by dougm — APPROVED** [review]
lgtm, thanks @embano1 !

**Pull request merged**